Method shape source generation results in ambiguity compile error

When a type inherits from two interfaces that each add a method with the same name and a similar signature, ambiguity may occur, which the source generated type shape provider does not address adequately.

```cs
[Fact]
public static void AmbiguousMethodInheritanceOnInterface()
{
    Compilation compilation = CompilationHelpers.CreateCompilation("""
        using PolyType;

        public interface IBase1
        {
            void DoSomething();
        }

        public interface IBase2
        {
            void DoSomething();
        }

        [GenerateShape(IncludeMethods = MethodShapeFlags.PublicInstance)]
        public partial interface IDerived : IBase1, IBase2
        {
            // This will cause an ambiguity error if not handled correctly
        }
        """);

    PolyTypeSourceGeneratorResult result = CompilationHelpers.RunPolyTypeSourceGenerator(compilation);
    Assert.Empty(result.Diagnostics);
}
```

## Expected

No error in generated code.
Both methods have shapes generated for them and may each be invoked. Ideally the shape would convey an idea of their origin so that an intelligent decision could be made, but I don't think I would use such a disambiguation, so that's optional.

## Actual

Generated code produces a compilation error:

> PolyType.SourceGenerator\PolyType.SourceGenerator.PolyTypeGenerator\PolyType.SourceGenerator.ShapeProvider.IDerived.g.cs(46,146): error CS0121: The call is ambiguous between the following methods or properties: 'IBase1.DoSomething()' and 'IBase2.DoSomething()'

Feel free to snag the commit that adds this test from [here](https://github.com/eiriktsarpalis/PolyType/compare/main...AArnott:PolyType:ambiguousMethodError).

## Suggestion

When invoking a method on some *object*, first assign the object to a local variable typed as the method's declaring type, then invoke the method on that local variable. This will *not* require a cast and will have no runtime perf hit. But it _will_ avoid the C# ambiguity error as it will generate IL that knows exactly which method was meant.
Alternatively, simply casting the target object to its declaring type and invoking the method directly on that cast expression will probably work and likely have no perf impact as well.

```diff
-MethodInvoker = static (ref global::IDerived? target, ref global::PolyType.SourceGenModel.EmptyArgumentState state) => { target!.DoSomething(); return new(global::PolyType.Abstractions.Unit.Value); },
+MethodInvoker = static (ref global::IDerived? target, ref global::PolyType.SourceGenModel.EmptyArgumentState state) => { IBase1 t = target!; t.DoSomething(); return new(global::PolyType.Abstractions.Unit.Value); },
```

## Justification

This came up during a StreamJsonRpc upgrade to utilize method shapes to enable RPC marshaled objects in a NativeAOT-safe environment. We have some tests that verified that StreamJsonRpc could deal with such a situation gracefully, and as soon as we enabled method shape generation for the derived type, compilation started failing.

--- a/src/PolyType.Examples/Cloner/Cloner.Builder.cs
+++ b/src/PolyType.Examples/Cloner/Cloner.Builder.cs
@@ -30,7 +30,7 @@ public static partial class Cloner
                 return CreatePolymorphicCloner(generationContext.ParentCache!);
             }
             
-            if (typeShape is { Properties: [], Constructor: null })
+            if (typeShape is { Type.IsValueType: true, Properties: [], Constructor: null })
             {
                 return new Func<T?, T?>(t => t);
             }

--- a/src/PolyType.Roslyn/Helpers/RoslynHelpers.cs
+++ b/src/PolyType.Roslyn/Helpers/RoslynHelpers.cs
@@ -348,9 +348,9 @@ internal static class RoslynHelpers
 
     public static ICollection<ITypeSymbol> GetSortedTypeHierarchy(this ITypeSymbol type)
     {
-        if (type.TypeKind != TypeKind.Interface)
+        if (type.TypeKind is not TypeKind.Interface)
         {
-            var list = new List<ITypeSymbol>();
+            List<ITypeSymbol> list = [];
             for (ITypeSymbol? current = type; current != null; current = current.BaseType)
             {
                 list.Add(current);
@@ -370,21 +370,39 @@ internal static class RoslynHelpers
     public static IEnumerable<ISymbol> GetAllMembers(this ITypeSymbol type)
         => type.GetSortedTypeHierarchy().SelectMany(t => t.GetMembers());
 
-    public static IEnumerable<IMethodSymbol> GetAllMethods(this ITypeSymbol type)
+    public static IEnumerable<(IMethodSymbol Symbol, bool IsAmbiguous)> GetAllMethods(this ITypeSymbol type)
     {
-        HashSet<string> seenMethods = new();
+        List<(IMethodSymbol Symbol, bool IsAmbiguous)> methods = [];
+        Dictionary<string, int> foundSignatures = new();
+
         foreach (ITypeSymbol current in type.GetSortedTypeHierarchy())
         {
             foreach (IMethodSymbol method in current.GetMembers().OfType<IMethodSymbol>())
             {
                 // Create a signature key that includes name, parameters, and return type
                 string signature = GetMethodSignature(method);
-                if (seenMethods.Add(signature))
+                if (foundSignatures.TryGetValue(signature, out int index))
                 {
-                    yield return method;
+                    IMethodSymbol conflictingMethod = methods[index].Symbol;
+                    if (!current.IsAssignableFrom(conflictingMethod.ContainingType))
+                    {
+                        // We have a method with the same signature in an unrelated type,
+                        // due to diamond ambiguity in an interface hierarchy.
+
+                        Debug.Assert(type.TypeKind is TypeKind.Interface);
+                        methods[index] = (conflictingMethod, IsAmbiguous: true);
+                        methods.Add((method, IsAmbiguous: true));
+                    }
+
+                    continue; // This method is shadowed by a derived type, skip it.
                 }
+
+                foundSignatures.Add(signature, methods.Count);
+                methods.Add((method, IsAmbiguous: false));
             }
         }
+
+        return methods;
 
         static string GetMethodSignature(IMethodSymbol method)
         {

--- a/src/PolyType.Roslyn/Model/MethodDataModel.cs
+++ b/src/PolyType.Roslyn/Model/MethodDataModel.cs
@@ -32,6 +32,11 @@ public readonly struct MethodDataModel
     /// The parameters of the method.
     /// </summary>
     public required ImmutableArray<ParameterDataModel> Parameters { get; init; }
+
+    /// <summary>
+    /// Whether the method is ambiguous due to diamond inheritance.
+    /// </summary>
+    public required bool IsDiamondAmbiguous { get; init; }
 }
 
 /// <summary>
@@ -68,4 +73,25 @@ public enum MethodReturnTypeKind
     /// The method returns a <see cref="System.Threading.Tasks.Task{TResult}"/>.
     /// </summary>
     TaskOfT,
+}
+
+/// <summary>
+/// A resolved method symbol that includes additional metadata.
+/// </summary>
+public readonly struct ResolvedMethodSymbol
+{
+    /// <summary>
+    /// Gets the resolved method symbol.
+    /// </summary>
+    public required IMethodSymbol MethodSymbol { get; init; }
+
+    /// <summary>
+    /// Gets a custom name to be applied to the method, if specified.
+    /// </summary>
+    public string? CustomName { get; init; }
+
+    /// <summary>
+    /// Gets a value indicating whether the method is ambiguous due to diamond inheritance.
+    /// </summary>
+    public bool IsDiamondAmbiguous { get; init; }
 }

--- a/src/PolyType.Roslyn/ModelGenerator/TypeDataModelGenerator.cs
+++ b/src/PolyType.Roslyn/ModelGenerator/TypeDataModelGenerator.cs
@@ -126,7 +126,7 @@ public partial class TypeDataModelGenerator
     /// <param name="type">The declaring type from which to resolve the methods.</param>
     /// <param name="bindingFlags">The binding flags of the methods to include.</param>
     /// <returns>A key/value pair containing the method and its declared name.</returns>
-    protected virtual IEnumerable<KeyValuePair<string, IMethodSymbol>> ResolveMethods(ITypeSymbol type, BindingFlags bindingFlags) => [];
+    protected virtual IEnumerable<ResolvedMethodSymbol> ResolveMethods(ITypeSymbol type, BindingFlags bindingFlags) => [];
 
     /// <summary>
     /// Wraps the <see cref="MapType(ITypeSymbol, TypeDataKind?, BindingFlags?, ImmutableArray{AssociatedTypeModel}, ref TypeDataModelGenerationContext, TypeShapeRequirements, out TypeDataModel?)"/> method
@@ -401,10 +401,10 @@ public partial class TypeDataModelGenerator
     protected ImmutableArray<MethodDataModel> MapMethods(ITypeSymbol type, ref TypeDataModelGenerationContext ctx, BindingFlags? bindingFlags)
     {
         ImmutableArray<MethodDataModel>.Builder results = ImmutableArray.CreateBuilder<MethodDataModel>();
-        foreach (KeyValuePair<string, IMethodSymbol> kvp in ResolveMethods(type, bindingFlags ?? BindingFlags.Default))
+        foreach (ResolvedMethodSymbol resolvedMethod in ResolveMethods(type, bindingFlags ?? BindingFlags.Default))
         {
             TypeDataModelGenerationContext scopedCtx = ctx;
-            if (MapMethod(kvp.Value, kvp.Key, ref scopedCtx, out MethodDataModel methodDataModel) is TypeDataModelGenerationStatus.Success)
+            if (MapMethod(resolvedMethod, ref scopedCtx, out MethodDataModel methodDataModel) is TypeDataModelGenerationStatus.Success)
             {
                 results.Add(methodDataModel);
                 ctx = scopedCtx;
@@ -415,19 +415,18 @@ public partial class TypeDataModelGenerator
     }
 
     /// <summary>
-    /// Maps the given <paramref name="method"/> and its parameters to a <see cref="MethodDataModel"/>.
+    /// Maps the given <paramref name="resolvedMethod"/> and its parameters to a <see cref="MethodDataModel"/>.
     /// </summary>
-    /// <param name="method">The method to try to map.</param>
-    /// <param name="name">The name identifying the method.</param>
+    /// <param name="resolvedMethod">The method to try to map.</param>
     /// <param name="ctx">The current model generation context.</param>
     /// <param name="result">The mapped method data model.</param>
     /// <returns>The result of the mapping operation.</returns>
-    protected virtual TypeDataModelGenerationStatus MapMethod(IMethodSymbol method, string name, ref TypeDataModelGenerationContext ctx, out MethodDataModel result)
+    protected virtual TypeDataModelGenerationStatus MapMethod(ResolvedMethodSymbol resolvedMethod, ref TypeDataModelGenerationContext ctx, out MethodDataModel result)
     {
         TypeDataModelGenerationStatus status;
 
-        var parameters = ImmutableArray.CreateBuilder<ParameterDataModel>(method.Parameters.Length);
-        foreach (IParameterSymbol parameter in method.Parameters)
+        var parameters = ImmutableArray.CreateBuilder<ParameterDataModel>(resolvedMethod.MethodSymbol.Parameters.Length);
+        foreach (IParameterSymbol parameter in resolvedMethod.MethodSymbol.Parameters)
         {
             if (parameter.RefKind is RefKind.Out)
             {
@@ -445,7 +444,7 @@ public partial class TypeDataModelGenerator
             parameters.Add(parameterModel);
         }
 
-        ITypeSymbol? returnType = GetEffectiveReturnType(method, out MethodReturnTypeKind returnTypeKind);
+        ITypeSymbol? returnType = GetEffectiveReturnType(resolvedMethod.MethodSymbol, out MethodReturnTypeKind returnTypeKind);
         if (returnType is not null && (status = IncludeNestedType(returnType, ref ctx)) != TypeDataModelGenerationStatus.Success)
         {
             result = default;
@@ -454,11 +453,12 @@ public partial class TypeDataModelGenerator
 
         result = new MethodDataModel
         {
-            Name = name,
-            Method = method,
+            Name = resolvedMethod.CustomName ?? resolvedMethod.MethodSymbol.Name,
+            Method = resolvedMethod.MethodSymbol,
             ReturnedValueType = returnType,
             ReturnTypeKind = returnTypeKind,
             Parameters = parameters.ToImmutable(),
+            IsDiamondAmbiguous = resolvedMethod.IsDiamondAmbiguous,
         };
 
         return TypeDataModelGenerationStatus.Success;

--- a/src/PolyType.SourceGenerator/Model/MethodShapeModel.cs
+++ b/src/PolyType.SourceGenerator/Model/MethodShapeModel.cs
@@ -15,6 +15,7 @@ public sealed record MethodShapeModel
     public required bool CanUseUnsafeAccessors { get; init; }
     public required bool IsStatic { get; init; }
     public required bool ReturnsByRef { get; init; }
+    public required bool InstanceMethodRequiresCast { get; init; }
     public required MethodReturnTypeKind ReturnTypeKind { get; init; }
     public required ImmutableEquatableArray<ParameterShapeModel> Parameters { get; init; }
     public required ArgumentStateType ArgumentStateType { get; init; }

--- a/src/PolyType.SourceGenerator/Parser/Parser.ModelMapper.cs
+++ b/src/PolyType.SourceGenerator/Parser/Parser.ModelMapper.cs
@@ -468,6 +468,7 @@ public sealed partial class Parser
                 UnderlyingReturnType = CreateTypeId(m.Method.ReturnType),
                 ReturnType = CreateTypeId(m.ReturnedValueType ?? _knownSymbols.UnitType!),
                 ReturnsByRef = m.Method.ReturnsByRef || m.Method.ReturnsByRefReadonly,
+                InstanceMethodRequiresCast = m.IsDiamondAmbiguous,
                 Parameters = m.Parameters
                     .Select(p => MapParameter(declaringObjectForConstructor: null, CreateTypeId(m.Method.ContainingType), p, false))
                     .ToImmutableEquatableArray(),

--- a/src/PolyType.SourceGenerator/Parser/Parser.cs
+++ b/src/PolyType.SourceGenerator/Parser/Parser.cs
@@ -304,13 +304,13 @@ public sealed partial class Parser : TypeDataModelGenerator
             .Take(1);
     }
 
-    protected override IEnumerable<KeyValuePair<string, IMethodSymbol>> ResolveMethods(ITypeSymbol type, BindingFlags bindingFlags)
+    protected override IEnumerable<ResolvedMethodSymbol> ResolveMethods(ITypeSymbol type, BindingFlags bindingFlags)
     {
-        foreach (IMethodSymbol method in type.GetAllMethods())
+        foreach ((IMethodSymbol method, bool isAmbiguous) in type.GetAllMethods())
         {
             if (IncludeMethod(method, out string? customName))
             {
-                yield return new(customName ?? method.Name, method);
+                yield return new() { CustomName = customName, MethodSymbol = method, IsDiamondAmbiguous = isAmbiguous };
             }
         }
 
@@ -377,19 +377,19 @@ public sealed partial class Parser : TypeDataModelGenerator
         }
     }
 
-    protected override TypeDataModelGenerationStatus MapMethod(IMethodSymbol method, string name, ref TypeDataModelGenerationContext ctx, out MethodDataModel result)
+    protected override TypeDataModelGenerationStatus MapMethod(ResolvedMethodSymbol resolvedMethod, ref TypeDataModelGenerationContext ctx, out MethodDataModel result)
     {
-        if (method is { IsGenericMethod: true, IsDefinition: true })
+        if (resolvedMethod.MethodSymbol is { IsGenericMethod: true, IsDefinition: true } method)
         {
             ReportDiagnostic(GenericMethodShapesNotSupported, method.Locations.FirstOrDefault(), [method.ToDisplayString()]);
             result = default;
             return TypeDataModelGenerationStatus.UnsupportedType;
         }
 
-        TypeDataModelGenerationStatus status = base.MapMethod(method, name, ref ctx, out result);
+        TypeDataModelGenerationStatus status = base.MapMethod(resolvedMethod, ref ctx, out result);
         if (status is not TypeDataModelGenerationStatus.Success)
         {
-            ReportDiagnostic(MethodParametersNotSupported, method.Locations.FirstOrDefault(), [method.ToDisplayString()]);
+            ReportDiagnostic(MethodParametersNotSupported, resolvedMethod.MethodSymbol.Locations.FirstOrDefault(), [resolvedMethod.MethodSymbol.ToDisplayString()]);
             return status;
         }
 

--- a/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Methods.cs
+++ b/src/PolyType.SourceGenerator/SourceFormatter/SourceFormatter.Methods.cs
@@ -158,13 +158,14 @@ internal sealed partial class SourceFormatter
             };
 
             string refPrefix = method.DeclaringType.IsValueType ? "ref " : "";
+            string targetExpr = method.InstanceMethodRequiresCast ? $"(({method.DeclaringType.FullyQualifiedName}){targetVar}!)" : $"{targetVar}!";
             string invokeExpr = method switch
             {
                 { IsStatic: true, IsAccessible: true } => $"{declaringType.Type.FullyQualifiedName}.{method.UnderlyingMethodName}({parametersExpression})",
                 { IsStatic: true, IsAccessible: false } => $"{GetMethodAccessorName(declaringType, method)}({parametersExpression})",
-                { IsStatic: false, IsAccessible: true } => $"{targetVar}!.{method.UnderlyingMethodName}({parametersExpression})",
-                { IsStatic: false, IsAccessible: false } when method.Parameters is [] => $"{GetMethodAccessorName(declaringType, method)}({refPrefix}{targetVar}!)",
-                { IsStatic: false, IsAccessible: false } => $"{GetMethodAccessorName(declaringType, method)}({refPrefix}{targetVar}!, {parametersExpression})",
+                { IsStatic: false, IsAccessible: true } => $"{targetExpr}.{method.UnderlyingMethodName}({parametersExpression})",
+                { IsStatic: false, IsAccessible: false } when method.Parameters is [] => $"{GetMethodAccessorName(declaringType, method)}({refPrefix}{targetExpr})",
+                { IsStatic: false, IsAccessible: false } => $"{GetMethodAccessorName(declaringType, method)}({refPrefix}{targetExpr}, {parametersExpression})",
             };
 
             // Handle async methods and void returns
@@ -325,7 +326,7 @@ internal sealed partial class SourceFormatter
         if (!method.IsStatic)
         {
             string refPrefix = method.DeclaringType.IsValueType ? "ref " : "";
-            parameterSignature.Append($"{refPrefix}{declaringType.Type.FullyQualifiedName} target, ");
+            parameterSignature.Append($"{refPrefix}{method.DeclaringType.FullyQualifiedName} target, ");
         }
 
         foreach (ParameterShapeModel parameter in method.Parameters)
@@ -360,7 +361,7 @@ internal sealed partial class SourceFormatter
                 private static global::System.Reflection.MethodInfo? __s_{{accessorName}}_MethodInfo;
                 private static {{method.UnderlyingReturnType.FullyQualifiedName}} {{accessorName}}({{allParameters}})
                 {
-                    global::System.Reflection.MethodInfo methodInfo = __s_{{accessorName}}_MethodInfo ??= typeof({{declaringType.Type.FullyQualifiedName}}).GetMethod({{FormatStringLiteral(method.UnderlyingMethodName)}}, {{AllBindingFlagsConstMember}}, null, {{parameterTypes}}, null)!;
+                    global::System.Reflection.MethodInfo methodInfo = __s_{{accessorName}}_MethodInfo ??= typeof({{method.DeclaringType}}).GetMethod({{FormatStringLiteral(method.UnderlyingMethodName)}}, {{AllBindingFlagsConstMember}}, null, {{parameterTypes}}, null)!;
                     object?[] paramArray = new object?[] { {{string.Join(", ", method.Parameters.Select(p => p.Name))}} };
                     {{(method.ReturnTypeKind is MethodReturnTypeKind.Void
                         ? CreateInvokeMethodExpr()

--- a/src/PolyType.TestCases/TestTypes.cs
+++ b/src/PolyType.TestCases/TestTypes.cs
@@ -629,6 +629,7 @@ public static class TestTypes
         yield return TestCase.Create<InterfaceWithMethodShapes>(new ClassWithMethodShapes(), additionalValues: [new StructWithMethodShapes()]);
         yield return TestCase.Create(new BaseClassWithMethodShapes(), additionalValues: [new ClassWithMethodShapes()]);
         yield return TestCase.Create(new RpcService());
+        yield return TestCase.Create<InterfaceWithDiamondMethodShapes>(new InterfaceWithDiamondMethodShapes.Impl());
     }
 
     private static ExpandoObject CreateExpandoObject(IEnumerable<KeyValuePair<string, object?>> values)
@@ -2930,6 +2931,28 @@ public partial class RpcService
     }
 
     public record Event(int id);
+}
+
+[GenerateShape, TypeShape(IncludeMethods = MethodShapeFlags.PublicInstance)]
+public partial interface InterfaceWithDiamondMethodShapes : IBase1WithMethod, IBase2WithMethod
+{
+    int Add3(int x, int y);
+
+    public sealed class Impl : InterfaceWithDiamondMethodShapes
+    {
+        public int Add(int x, int y) => x + y;
+        public int Add3(int x, int y) => x + y;
+    }
+}
+
+public interface IBase1WithMethod
+{
+    int Add(int x, int y);
+}
+
+public interface IBase2WithMethod
+{
+    int Add(int x, int y);
 }
 
 [GenerateShapeFor<object>]

--- a/tests/PolyType.SourceGenerator.UnitTests/CompilationTests.MethodShapes.cs
+++ b/tests/PolyType.SourceGenerator.UnitTests/CompilationTests.MethodShapes.cs
@@ -406,4 +406,30 @@ public static partial class CompilationTests
         PolyTypeSourceGeneratorResult result = CompilationHelpers.RunPolyTypeSourceGenerator(compilation);
         Assert.Empty(result.Diagnostics);
     }
+
+    [Fact]
+    public static void MethodDiamondAmbiguity_NoErrors()
+    {
+        Compilation compilation = CompilationHelpers.CreateCompilation("""
+        using PolyType;
+
+        public interface IBase1
+        {
+            void DoSomething();
+        }
+
+        public interface IBase2
+        {
+            void DoSomething();
+        }
+
+        [GenerateShape(IncludeMethods = MethodShapeFlags.PublicInstance)]
+        public partial interface IDerived : IBase1, IBase2
+        {
+        }
+        """);
+
+        PolyTypeSourceGeneratorResult result = CompilationHelpers.RunPolyTypeSourceGenerator(compilation);
+        Assert.Empty(result.Diagnostics);
+    }
 }

--- a/tests/PolyType.Tests/ClonerTests.cs
+++ b/tests/PolyType.Tests/ClonerTests.cs
@@ -13,7 +13,7 @@ public abstract class ClonerTests(ProviderUnderTest providerUnderTest)
     {
         (Func<T?, T?> cloner, IEqualityComparer<T> comparer) = GetClonerAndEqualityComparer(testCase);
 
-        if (testCase.Value is not null && !providerUnderTest.HasConstructor(testCase) && testCase.CustomKind is not TypeShapeKind.None)
+        if (testCase.Value is not null && !providerUnderTest.HasConstructor(testCase))
         {
             Assert.Throws<NotSupportedException>(() => cloner(testCase.Value));
             return;

